### PR TITLE
Fix unified parallel h5 output

### LINKED ISSUE #581: Error message from multiple chains writing HDF5: `hdf file type not supported for multiple sub-environments yet`
Input file:

```
$ cat queso_test.inp8
env_numSubEnvironments   = 8
env_subDisplayFileName   = .
env_subDisplayAllowAll   = 0
env_displayVerbosity     = 0
env_syncVerbosity        = 0
env_seed                 = 0

ip_computeSolution      = 1
ip_dataOutputFileName   = .

ip_mh_dataOutputFileName   = .
ip_mh_dataOutputAllowAll   = 0

ip_mh_rawChain_dataInputFileName    = .
ip_mh_rawChain_size                 = 100000
ip_mh_rawChain_generateExtra        = 0
ip_mh_rawChain_dataOutputFileName   = outputData8/ip_raw_chain
ip_mh_rawChain_dataOutputFileType   = h5
ip_mh_rawChain_dataOutputAllowedSet = 0 1
ip_mh_rawChain_computeStats         = 1

ip_mh_displayCandidates             = 0

ip_mh_filteredChain_generate             = 0
```

Here's what QUESO wrote to the screen:

```
../../src/basic/src/SequenceOfVectors.C, line 1687, compiled May 22 2017 at 13:40:38
hdf file type not supported for multiple sub-environments yet
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 1

===================================================================================
=   BAD TERMINATION OF ONE OF YOUR APPLICATION PROCESSES
=   PID 3355 RUNNING AT zoidberg.ices.utexas.edu
=   EXIT CODE: 1
=   CLEANING UP REMAINING PROCESSES
=   YOU CAN IGNORE THE BELOW CLEANUP MESSAGES
===================================================================================
```

Here's my output directory:

```
$ ll outputData8/
total 2.4M
-rw-r--r-- 1 damon pecos  800 May 22 14:39 ip_raw_chain.h5
-rw-r--r-- 1 damon pecos 784K May 22 14:39 ip_raw_chain_loglikelihood_sub0.h5
-rw-r--r-- 1 damon pecos 784K May 22 14:39 ip_raw_chain_logtarget_sub0.h5
-rw-r--r-- 1 damon pecos 784K May 22 14:39 ip_raw_chain_sub0.h5

$ h5dump outputData8/ip_raw_chain.h5
HDF5 "outputData8/ip_raw_chain.h5" {
GROUP "/" {
}
}

$ h5dump outputData8/ip_raw_chain_sub0.h5 | head
HDF5 "outputData8/ip_raw_chain_sub0.h5" {
GROUP "/" {
   DATASET "data" {
      DATATYPE  H5T_IEEE_F64LE
      DATASPACE  SIMPLE { ( 100000, 1 ) / ( 100000, 1 ) }
      DATA {
      (0,0): 6,
      (1,0): 6.07047,
      (2,0): 6.07047,
      (3,0): 5.97638,
```

The specific-process output chains are written, but not the concatenated one.

I suspect this is just a logic bug.

### PR DESCRIPTION
**Pull request by dmcdougall**
Fixes #581.

### CODE REVIEW

File: src/core/src/Environment.C [hunk 1]
```diff
@@ -842,14 +850,20 @@ BaseEnvironment::openUnifiedOutputFile(
         }
 #ifdef QUESO_HAS_HDF5
         else if (fileType == UQ_FILE_EXTENSION_FOR_HDF_FORMAT) {
-          filePtrSet.h5Var = H5Fcreate((baseFileName+"."+fileType).c_str(), // TEMPORARY - FIX ME
+
+          std::string fullFileName = baseFileName+"."+fileType;
+
+          filePtrSet.h5Var = H5Fcreate(fullFileName.c_str(), // TEMPORARY - FIX ME
```
Comment (roystgnr) [inline src/core/src/Environment.C]: This comment should either get a removal or an explanation

File: test/test_StatisticalInverseProblem/test_parallel_h5.sh.in [hunk 2]
```diff
@@ -0,0 +1,22 @@
+#!/bin/bash
+set -eu
+set -o pipefail
+
+have_mpi=@HAVE_MPI@
+have_hdf5=@HAVE_HDF5@
+
+if [ $have_mpi -eq 1 ]; then
+  if [ $have_hdf5 -eq 1 ]; then
+    PROG="mpirun -np 2 ./test_parallel_h5"
```
Comment (roystgnr) [inline test/test_StatisticalInverseProblem/test_parallel_h5.sh.in]: Maybe use mpiexec here?  I'm not sure which is more common but IIRC mpiexec is the only one that appears (albeit as optional to support...) in an actual MPI standard.

File: test/test_StatisticalInverseProblem/test_parallel_h5.sh.in [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+    INPUT="${srcdir}/test_StatisticalInverseProblem/input_test_parallel_h5.txt"
+
+    $PROG $INPUT
+
```
Comment (roystgnr) [inline test/test_StatisticalInverseProblem/test_parallel_h5.sh.in:15]: Would it be possible to use h5diff, h5ls, or ls (in descending order of preference) to make sure the output is sane before we trash it?

File: src/basic/src/SequenceOfVectors.C [hunk 4]
```diff
@@ -1536,6 +1538,93 @@ SequenceOfVectors<V,M>::unifiedWriteContents(
...
+
+      m_env.inter0Comm().template Gather<double>(data,
+                                                 numParams * chainSize,
+                                                 recvbuf,
+                                                 numParams * chainSize,
+                                                 0,
+                                                 "",
+                                                 "");
+
+      if (m_env.inter0Rank() == 0) {
```
Comment (roystgnr) [inline src/basic/src/SequenceOfVectors.C:1566]: This looks like the right thing to do for now, but in the long run phdf5 support would be nice.

File: src/basic/src/SequenceOfVectors.C [hunk 5]
```diff
@@ -1536,6 +1538,93 @@ SequenceOfVectors<V,M>::unifiedWriteContents(
   }
 
   if (m_env.inter0Rank() >= 0) {
+    if (fileType == UQ_FILE_EXTENSION_FOR_HDF_FORMAT) {
+#ifdef QUESO_HAS_HDF5
+      unsigned int chainSize = this->subSequenceSize();
+      unsigned int numParams = m_vectorSpace.dimLocal();
+      double * data;
+      data = (double *)malloc(numParams * chainSize * sizeof(double));
```
Comment (roystgnr) [inline src/basic/src/SequenceOfVectors.C]: std::vector, instead, for the exception safety?

File: src/basic/src/ScalarSequence.C [hunk 6]
```diff
@@ -2685,6 +2685,82 @@ ScalarSequence<T>::unifiedWriteContents(
   // As of 14/Nov/2009, this routine does *not* require sub sequences to have equal size. Good.
 
   if (m_env.inter0Rank() >= 0) {
+    if (fileType == UQ_FILE_EXTENSION_FOR_HDF_FORMAT) {
+#ifdef QUESO_HAS_HDF5
+      unsigned int chainSize = this->subSequenceSize();
+      unsigned int numChains = m_env.inter0Comm().NumProc();
+
+      double * recvbuf;
+      recvbuf = (double *)malloc(chainSize * numChains * sizeof(double));
```
Comment (roystgnr) [inline src/basic/src/ScalarSequence.C:2694]: Likewise std::vector would be safer here.

**Comment by roystgnr** [comment]
:+1: to merge regardless, but the more of my wishlist you hit first the happier I'll be afterwards.

File: src/basic/src/SequenceOfVectors.C (hunk 4 above)
Comment (dmcdougall) [reply src/basic/src/SequenceOfVectors.C:1566]: Agreed.

**Pull request merged**